### PR TITLE
Additive and Multiplicative

### LINKED ISSUE #8: Add `Additive` and `Multiplicative` monoids
Should we release as `v0.2.0` after this, or just `v0.1.6`? The latter has the advantage of us not having to rerelease that massive list of things I mentioned in the issue about `Num`, but as this will require a new version of the compiler it sort of is a breaking change.

### PR DESCRIPTION
**Pull request by garyb**
@paf31 / @jdegoes to review please. Also are there any instances worth having that I'm missing here?

Resolves #8

### CODE REVIEW

File: src/Data/Monoid/Additive.purs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+module Data.Monoid.Additive where
```
Comment (jdegoes) [inline src/Data/Monoid/Additive.purs]: If you really want to go crazy, you can also define `Alternative`, `MonadPlus`.

There's also `Foldable` and `Traversable`.

Same for `Multiplicative`.

**Comment by jdegoes** [comment]
I think `Traversable` is the only one I could see myself using (to save a few chars). Others listed above.

**Comment by jdegoes** [comment]
Looks great, btw! :ship:

**Comment by garyb** [comment]
I tried defining `Alternative` actually, but the compiler didn't seem to like it. User error perhaps, I'll try again. Thanks!

**Comment by garyb** [comment]
I remember now... since there is no way of introducing an `Alternative a` implication for `Additive a`, I don't think it's possible to define `Alternative`, is it?

**Comment by garyb** [comment]
I'll define the `Foldable` and `Traversable` instances in `purescript-foldable-traversable` as that has a dependency on `purescript-monoid`

**Comment by jdegoes** [comment]
`Alternative` just requires `Plus`, which requires `empty` and `Alt`. The empties would be zero for `Additive` and one for `Multiplicative`, while `(<|>)` would  be addition / multiplication, respectively. I believe these definitions are consistent with all laws.

**Comment by garyb** [comment]
Whoops, I was being a bit dense before. However, this formulation has a similar problem since the contents of `Additive` is `a`, there is no `Semiring` implication to allow me to `+` them or use `zero`.

**Comment by jdegoes** [comment]
Can you do:

```
instance altAdditive :: (Semiring a) => Alt Additive where
  (<|>) (Additive l) (Additive r) = Additive (l + r)

instance plusAdditive :: (Semiring a) => Plus Additive where
  empty = zero

instance monadPlusAdditive :: ...
```

???

**Comment by garyb** [comment]
No, the `a` would have to appear on the right hand side too for that, but then we'd have the wrong kind.

**Comment by jdegoes** [comment]
Ah yes, of course, these constraints would have to be applied polymorphically, which is not possible...

**Comment by paf31** [comment]
I think the instances provided should reflect the usage of the module. A beginner should be able to understand the intent of `Additive` by reading the instances it provides. 

What is `Additive` for? It provides a `Monoid` instance which uses `(+)` from `Semiring`. I think providing things like `Functor` confuses things to a certain extent. An value of type `Additive a` doesn't make sense if `a` is not a `Semiring`.

**Comment by garyb** [comment]
I guess the argument for having them is if we don't define those instances here then they can never be defined without orphan instances or even more `newtype`ing.

**Comment by jdegoes** [comment]
Well, let's say there's a library-level function `f` which operates on `Number`, and you want to lift it to `Additive` because you're in a bunch of code requiring the `Additive` monoid. It's nice to be able to do things like, `f <$> x <*> y`. A broader point is that, to avoid orphan instances, if *anyone, ever* wants to be able to use these instances, they have to be defined here.

**Comment by paf31** [comment]
Makes sense.

**Comment by garyb** [comment]
Yeah, I think that would be a good way to do it. I should have added some while I was working on this actually, will do so and then merge after that's reviewed also.

**Comment by garyb** [comment]
Does this make `Sum` and `Product` redundant then? Should we just update those to use `Semiring` instead?

**Comment by garyb** [comment]
Any thoughts on `Sum`/`Product` @paf31, @jdegoes?

**Comment by jdegoes** [comment]
Makes sense to reuse those, add the new instances.

**Comment by paf31** [comment]
I think we can either keep them or remove them, but I'd rather not turn `Sum` into `Additive`.

**Comment by garyb** [comment]
Are you talking about names there, or do you mean you'd prefer to keep `Sum` as being `Number`-specialised?

**Comment by paf31** [comment]
I'm saying that if we keep `Sum`, it should still be specialized to `Number`, instead of changing its behavior. I liken this situation to `ErrorT` vs `ExceptT`.

**Comment by garyb** [comment]
I'd vote to remove them then, as generally we've been avoiding monomorphic types if there is a polymorphic alternative, unless it's essential for performance.

**Comment by garyb** [comment]
This release will be a "breaking change version" too so should have no immediate impact if there are people using `Sum` and `Product`.

**Comment by jdegoes** [comment]
:+1: to removing the monomorphic versions.

**Comment by paf31** [comment]
Sounds good to me.

**Comment by paf31** [comment]
I haven't released the compiler yet though, so this will break until then. Can we hold off on the release for a day or two until I can push to Hackage?

**Comment by garyb** [comment]
Yep, no problem.

I'm going to make a PR with a few other instances and other bits for the other monoid newtypes yet too.

**Pull request merged**